src/DataStructures/VectorImpl.hpp

```diff
@@ -552,3 +552,46 @@ std::ostream& operator<<(std::ostream& os,
     }                                                    \
   };                                                     \
   }  // namespace MakeWithValueImpls
+
+// {@
+/*!
+ * \ingroup DataStructuresGroup
+ * \ingroup TypeTraitsGroup
+ * \brief Helper struct to determine the base storage type of a VectorImpl or
+ * container of VectorImpl
+ *
+ * \details Extracts the storage type of a `VectorImpl`, a std::array of
+ * `VectorImpl`, or a reference or pointer to a `VectorImpl`. In any of these
+ * cases, the `type` member is defined as the `ElementType` of the `VectorImpl`
+ * in question. If, instead, `vector_base_type` is passed a numeric type, the
+ * `type` member is defined as that numeric type.
+ *
+ * \snippet DataStructures/Test_VectorImpl.cpp vector_base_type_example
+ */
+template <typename T, typename ValidVectorElement =
+                          cpp17::bool_constant<cpp17::is_arithmetic_v<T> or
```

Comment:
I don't think you need the `bool_constant`, just replace `typename ValidVectorElement` with `bool` instead (I don't think you even need to name it)
```cpp
bool = cpp17::is_arithmetic_v<T> or tt::is_a_v<std::complex, T>
```

Reply:
turns out that doesn't work, it's what I tried first. Non-type template arguments aren't permitted to depend on the partial template specialization arguments.

Reply:
I'm not so sure...
```cpp
class DataVector {
 public:
  using ResultType = DataVector;
  using ElementType = double;
};

template <template <typename...> class U, typename T>
struct is_a : std::false_type {};
/// \cond HIDDEN_SYMBOLS
template <template <typename...> class U, typename... Args>
struct is_a<U, U<Args...>> : std::true_type {};
/// \endcond
/// \see is_a
template <template <typename...> class U, typename... Args>
constexpr bool is_a_v = is_a<U, Args...>::value;

/// \see is_a
template <template <typename...> class U, typename... Args>
using is_a_t = typename is_a<U, Args...>::type;

template <typename T,
          bool = std::is_arithmetic<T>::value or is_a_v<std::complex, T>>
struct vector_base_type {
  using type = T;
};
template <typename T>
struct vector_base_type<T, false> {
  using type =
      typename vector_base_type<typename T::ResultType::ElementType>::type;
};
template <typename T>
struct vector_base_type<T*, false> {
  using type = typename vector_base_type<T>::type;
};
template <typename T>
struct vector_base_type<T&, false> {
  using type = typename vector_base_type<T>::type;
};
template <typename T, size_t S>
struct vector_base_type<std::array<T, S>, false> {
  using type = typename vector_base_type<T>::type;
};

template <typename T>
using vector_base_type_t = typename vector_base_type<T>::type;

static_assert(std::is_same<vector_base_type_t<DataVector>, double>::value,
              "Failed testing type trait vector_base_type");
static_assert(
    std::is_same<vector_base_type_t<std::array<DataVector, 2>>, double>::value,
    "Failed testing type trait vector_base_type");
static_assert(std::is_same<vector_base_type_t<std::complex<double>*>,
                           std::complex<double>>::value,
              "Failed testing type trait vector_base_type");
static_assert(std::is_same<vector_base_type_t<DataVector&>, double>::value,
              "Failed testing type trait vector_base_type");
```
compiles on Clang7, GCC6, GCC7, and GCC8

Reply:
hmm... you've got me. I must have made the mistake of putting the `is_arithmetic` test in the specialization the first time I tried it. Alright, I'll put in this change and an additional alteration to make it compiler-error when a class that isn't a vector or contained type is passed.